File: forge/features/middleware/templates/rate_limit/python/files/src/app/middleware/rate_limit.py
"""In-memory per-tenant/per-IP token-bucket rate limiter."""

from __future__ import annotations

import ipaddress
import logging
import time
from collections import OrderedDict
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field

from fastapi import Request, Response, status
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware

logger = logging.getLogger(__name__)


def is_trusted_peer(ip: str) -> bool:
    """Return ``True`` when ``ip`` is an in-cluster / proxy peer we trust to
    have set ``X-Forwarded-For`` honestly.

    Heuristic: trust loopback, RFC1918 private, and link-local addresses — the
    documented Traefik topology terminates external traffic at an in-cluster
    reverse proxy whose peer address is private. A public transport peer means
    the request reached us directly (or via an XFF-appending hop), so a
    client-supplied XFF can't be trusted. A stricter explicit CIDR allowlist
    can be layered on top of this later.
    """
    try:
        addr = ipaddress.ip_address(ip)
    except ValueError:
        return False
    return addr.is_private or addr.is_loopback or addr.is_link_local

# Cap on the number of distinct client buckets kept in memory. Once exceeded,
# the least-recently-used bucket is evicted. Without this, a flood of unique
# clients (each a fresh X-Forwarded-For address) would grow the map without
# bound — a memory-exhaustion vector.
_MAX_BUCKETS = 4096


@dataclass
class _Bucket:
    tokens: float
    last_refill: float = field(default_factory=time.monotonic)


class RateLimitMiddleware(BaseHTTPMiddleware):
    def __init__(
        self,
        app,
        *,
        requests_per_minute: int = 120,
        burst: int | None = None,
        skip_paths: list[str] | None = None,
    ):
        super().__init__(app)
        self._rate = requests_per_minute / 60.0
        self._capacity = float(burst or requests_per_minute)
        # LRU-ordered so we can evict the least-recently-used idle bucket once
        # the map is full, keeping the limiter's memory bounded.
        self._buckets: OrderedDict[str, _Bucket] = OrderedDict()
        self._skip_paths = set(skip_paths or [])

    def _get_bucket(self, key: str) -> _Bucket:
        bucket = self._buckets.get(key)
        if bucket is None:
            # Evict the oldest (least-recently-used) bucket(s) before inserting
            # so the map can never exceed the cap.
            while len(self._buckets) >= _MAX_BUCKETS:
                self._buckets.popitem(last=False)
            bucket = _Bucket(tokens=self._capacity)
            self._buckets[key] = bucket
        else:
            # Mark as most-recently-used.
            self._buckets.move_to_end(key)
        return bucket

    async def dispatch(
        self, request: Request, call_next: Callable[[Request], Awaitable[Response]]
    ) -> Response:
        if any(request.url.path.startswith(p) for p in self._skip_paths):
            return await call_next(request)

        key = self._resolve_key(request)
        bucket = self._get_bucket(key)
        now = time.monotonic()

        elapsed = now - bucket.last_refill
        bucket.tokens = min(self._capacity, bucket.tokens + elapsed * self._rate)
        bucket.last_refill = now

        if bucket.tokens < 1.0:
            retry_after = int((1.0 - bucket.tokens) / self._rate) + 1
            logger.warning("Rate limit exceeded for key=%s", key)
            return JSONResponse(
                status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                content={"detail": "Rate limit exceeded. Please slow down."},
                headers={"Retry-After": str(retry_after)},
            )

        bucket.tokens -= 1.0
        return await call_next(request)

    @staticmethod
    def _resolve_key(request: Request) -> str:
        user = getattr(request.state, "user", None)
        if user is not None:
            customer_id = getattr(user, "customer_id", None)
            if customer_id:
                return f"tenant:{customer_id}"
        # Behind a reverse proxy / load balancer, ``request.client.host`` is the
        # proxy's address, shared by every anonymous client — keying on it would
        # collapse them all into one bucket. Prefer the left-most (originating)
        # address from X-Forwarded-For when present.
        #
        # SECURITY: only trust X-Forwarded-For when the immediate transport peer
        # is itself a trusted in-cluster proxy (private / loopback / link-local
        # — the documented Traefik topology). If the peer is a public address
        # the request reached us directly (internet-facing) or via an
        # XFF-appending hop, and a client-supplied XFF could be used to spoof /
        # evade its rate-limit bucket — so we ignore it and key on the peer.
        peer_host = request.client.host if request.client else None
        if peer_host and is_trusted_peer(peer_host):
            forwarded = request.headers.get("x-forwarded-for")
            if forwarded:
                client_ip = forwarded.split(",", 1)[0].strip()
                if client_ip:
                    return f"ip:{client_ip}"
        if peer_host:
            return f"ip:{peer_host}"
        return "anonymous"

File: forge/features/middleware/templates/rate_limit/rust/files/src/middleware/rate_limit.rs
//! Per-client token-bucket rate limiter.
//!
//! Keys on the originating client: the left-most X-Forwarded-For address when
//! present (so distinct clients behind a proxy don't share one bucket), else
//! the ConnectInfo transport peer. The bucket map is bounded (idle buckets are
//! evicted) so unique-client floods can't exhaust memory.
//!
//! In-memory only — fine for a single-instance deployment, unsuitable for
//! a horizontally-scaled stack (each replica maintains its own buckets).
//! Swap for a Redis-backed implementation if you deploy multiple replicas.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::Instant;

use axum::body::Body;
use axum::extract::ConnectInfo;
use axum::http::{Request, StatusCode};
use axum::middleware::Next;
use axum::response::{IntoResponse, Response};
use std::net::SocketAddr;

const REQUESTS_PER_MINUTE: f64 = 120.0;
const BURST: f64 = 120.0;
/// Cap on the number of distinct client buckets kept in memory. Once the map
/// grows past this, the least-recently-seen idle buckets are evicted. Without
/// this bound a flood of unique clients (each a fresh X-Forwarded-For address)
/// would grow the map without limit — a memory-exhaustion vector.
const MAX_BUCKETS: usize = 4096;

#[derive(Clone)]
pub struct RateLimiter {
    buckets: Arc<Mutex<HashMap<String, Bucket>>>,
    rate_per_sec: f64,
    capacity: f64,
}

struct Bucket {
    tokens: f64,
    last_refill: Instant,
}

impl RateLimiter {
    pub fn new() -> Self {
        Self {
            buckets: Arc::new(Mutex::new(HashMap::new())),
            rate_per_sec: REQUESTS_PER_MINUTE / 60.0,
            capacity: BURST,
        }
    }

    fn check(&self, key: &str) -> bool {
        let mut buckets = self.buckets.lock().unwrap();
        let now = Instant::now();

        // Bound the map: if we're at capacity and this is a new key, evict
        // buckets that have been full (idle) long enough to have fully
        // refilled — they carry no state worth keeping. This keeps the map
        // from growing without bound under a flood of unique clients.
        if buckets.len() >= MAX_BUCKETS && !buckets.contains_key(key) {
            let capacity = self.capacity;
            let rate = self.rate_per_sec;
            buckets.retain(|_, b| {
                let refilled = (b.tokens + now.duration_since(b.last_refill).as_secs_f64() * rate)
                    .min(capacity);
                // Keep buckets that still owe tokens (actively limited); drop
                // ones that have fully refilled and are effectively idle.
                refilled < capacity
            });
            // If every bucket is still active (the idle retain freed nothing),
            // the map is still at the cap. Hard-bound it by evicting the
            // least-recently-used bucket (smallest last_refill) before inserting,
            // mirroring the Python limiter's OrderedDict.popitem hard cap. Without
            // this, a flood of unique active clients would grow the map without
            // limit.
            if buckets.len() >= MAX_BUCKETS {
                if let Some(lru_key) = buckets
                    .iter()
                    .min_by(|a, b| a.1.last_refill.cmp(&b.1.last_refill))
                    .map(|(k, _)| k.clone())
                {
                    buckets.remove(&lru_key);
                }
            }
        }

        let bucket = buckets.entry(key.to_string()).or_insert_with(|| Bucket {
            tokens: self.capacity,
            last_refill: now,
        });
        let elapsed = now.duration_since(bucket.last_refill).as_secs_f64();
        bucket.tokens = (bucket.tokens + elapsed * self.rate_per_sec).min(self.capacity);
        bucket.last_refill = now;
        if bucket.tokens < 1.0 {
            return false;
        }
        bucket.tokens -= 1.0;
        true
    }
}

impl Default for RateLimiter {
    fn default() -> Self {
        Self::new()
    }
}

pub async fn rate_limit_middleware(req: Request<Body>, next: Next) -> Response {
    // Skip health/metrics paths so probes aren't rate-limited.
    let path = req.uri().path();
    if path.starts_with("/health") || path.starts_with("/metrics") {
        return next.run(req).await;
    }

    // Behind a reverse proxy / load balancer the ConnectInfo peer is the
    // proxy's address, shared by every anonymous client — keying on it would
    // collapse them all into one bucket. Prefer the left-most (originating)
    // address from X-Forwarded-For when present.
    //
    // NOTE: only trust X-Forwarded-For when the app sits behind a trusted proxy
    // that sets it; if clients can reach the app directly, strip/override the
    // header at the proxy so it can't be spoofed.
    let forwarded_ip = req
        .headers()
        .get("x-forwarded-for")
        .and_then(|value| value.to_str().ok())
        .and_then(|value| value.split(',').next())
        .map(str::trim)
        .filter(|ip| !ip.is_empty())
        .map(str::to_string);

    let key = forwarded_ip
        .or_else(|| {
            // Fall back to the ConnectInfo extension (set by axum::serve
            // with_connect_info) when there's no trusted forwarded header.
            req.extensions()
                .get::<ConnectInfo<SocketAddr>>()
                .map(|ConnectInfo(addr)| addr.ip().to_string())
        })
        .unwrap_or_else(|| "anonymous".to_string());

    // Use a per-process singleton limiter so state persists across requests.
    static LIMITER: std::sync::OnceLock<RateLimiter> = std::sync::OnceLock::new();
    let limiter = LIMITER.get_or_init(RateLimiter::new);

    if !limiter.check(&key) {
        return (
            StatusCode::TOO_MANY_REQUESTS,
            [("retry-after", "60")],
            "Rate limit exceeded. Please slow down.",
        )
            .into_response();
    }
    next.run(req).await
}
